design/external-authorization-design.md

```diff
@@ -0,0 +1,453 @@
+This means that the scope of how Contour can configure authorizationi depends on the structure of the HTTP configuration.
+For HTTP, there is only one HTTP Connection Manager that contains all the virtual hosts.
+Configuring a single authorization service for all the virtual hosts might be acceptable for some organizations, but it doesn't really fix the Contour multi-tenancy model.
+On the other hand, Contour configures HTTPS with a separate HTTP Connection Manager for each virtual host, so different authorization servers can naturally be attached to each HTTPS virtual host.
+
+The result of this is that authorization servers are not supported on HTTP.
+While it is convenient from an implementation perspective, this policy is also consistent with Contour's security-first posture.
+Note that using the TLS fallback certificate (for non-SNI clients) has the same HTTP Connection Manager properties as HTTP, so authentication servers also cannot be configured on virtual hosts that share the fallback certificate.
+
+## Detailed Design
+
+### SupportService Changes
+
+The `SupportService` CRD has API version `projectcontour.io/v1alpha1`.
+
+```Go
+// SupportProtocolVersion is the version of the GRPC protocol used
+// to request support services.
+type SupportProtocolVersion string
+
+// SupportProtocolVersion2 requests the "v2" support protocol version.
+const SupportProtocolVersion2 SupportProtocolVersion = "v2"
+
+type SupportService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   SupportServiceSpec   `json:"spec,omitempty"`
+	Status SupportServiceStatus `json:"status,omitempty"`
+}
+
+type SupportServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
+}
+
+type SupportServiceSpec struct {
+	// +required
+	Service contourv1.Service `json:"service"`
+
+	// The load balancing policy for sending requests to the service.
+	// +optional
+	LoadBalancerPolicy *contourv1.LoadBalancerPolicy `json:"loadBalancerPolicy,omitempty"`
+
+	// This field sets the version of the GRPC protocol that Envoy uses to
+	// send requests to the support service. Since Contour always uses the
+	// v2 Envoy API, this is currently fixed at "v2". However, other
+	// protocol options will be available in future.
+	//
+	// +kubebuilder:validation:Enum=v2
+	ProtocolVersion SupportProtocolVersion `json:"protocolVersion,omitempty"`
+}
+```
+
+The `Conditions` field in `SupportServicesStatus` follows the standardized upstream structure.
+This type should be shared with other Contour API status types.
+
+This document does not yet define how Contour monitors the support service.
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should monitor the underlying Kubernetes service and expose whether there are sufficient healthy endpoints.
```

Comment:
Ideally we encourage folks to setup readiness probes and rely on k8s to do the lifting, Contour is just going to record which of those pass the probe. My point here is that Contour shouldn't be doing active checks against the external service.

Reply:
Previously, when dependencies on other services have been proposed, having Contour make active availability checks against those has been suggested. Here, I'm not saying we should actually do that, but it's a possibility that we can leave open.

What I'm suggesting here is consistent with your suggestion I think. Rather than operators checking the state of the underlying service pods, Contour can do that on their behalf (since it already knows about endpoints), and roll that information up to the support service status.

Reply:
> What I'm suggesting here is consistent with your suggestion I think.

This was not my intent, my thought was to take Contour out of the business of doing these checks. They are external systems that Contour doesn't control, we should use the primitives provided by Kubernetes to implement.

Reply:
> > What I'm suggesting here is consistent with your suggestion I think.
> 
> This was not my intent, my thought was to take Contour out of the business of doing these checks. They are external systems that Contour doesn't control, we should use the primitives provided by Kubernetes to implement.

Yes, I agree with you :)

what I'm trying to say is that "monitor the underlying Kubernetes service" doesn't have to mean "actively health check"; it can mean "see whether there are any endpoints alive", or whatever we agree makes sense. The goal here is that we can take advantage of the CRD status to make the status of the underlying objects more obvious, just like deployments do for pods.

https://github.com/projectcontour/contour/issues/2325

Reply:
Maybe something like `Contour should observe the status of the service, directly or indirectly` would make this clearer.